packages/app/src/debug/DebugContainer.vue

```diff
@@ -0,0 +1,100 @@
+<template>
+  <div>
+    <div
+      v-if="loginConnectStore.user.isLoggedIn && loginConnectStore.project.isProjectConnected && run"
+    >
+      <DebugPageHeader
+        :gql="run"
+        :commits-ahead="0"
+      />
+      <DebugSpecList
+        :specs="debugSpecsArray"
+      />
+    </div>
+    <div
+      data-cy="debug-empty"
+    >
+      <div
+        v-if="!loginConnectStore.user.isLoggedIn"
+      >
+        {{ t('debugPage.notLoggedIn') }}
+      </div>
+      <div
+        v-if="loginConnectStore.user.isLoggedIn && !loginConnectStore.project.isProjectConnected"
+      >
+        {{ t('debugPage.notConnected' ) }}
+      </div>
+      <div
+        v-if="loginConnectStore.user.isLoggedIn && loginConnectStore.project.isProjectConnected && !run"
+      >
+        {{ t('debugPage.noRuns') }}
+      </div>
+    </div>
+    <!-- <pre>{{ JSON.stringify(props.gql, null, 2) }}</pre> -->
+  </div>
+</template>
+
+<script setup lang="ts">
+import { gql } from '@urql/vue'
+import { computed } from '@vue/reactivity'
+import type { DebugSpecsFragment } from '../generated/graphql'
+import { useLoginConnectStore } from '@packages/frontend-shared/src/store/login-connect-store'
+import DebugPageHeader from './DebugPageHeader.vue'
+import DebugSpecList from './DebugSpecList.vue'
+import { useI18n } from 'vue-i18n'
+import { specsList } from './utils/DebugMapping'
+
+const { t } = useI18n()
+
+gql`
+  fragment DebugSpecs on Query {
+    currentProject {
+      id
+      cloudProject {
+        __typename
+        ... on CloudProject {
+          id
+          runByNumber(runNumber: 2) {
+            ...DebugPage
+            id
+            runNumber
+            status
+            overLimitActionType
+            overLimitActionUrl
+            testsForReview {
+              id
+              ...DebugSpecListTests
+            }
+            specs {
+              id
+              ...DebugSpecListSpec
+            }
+          }
+        }
+      }
+    }
+  }
+`
+
+const props = defineProps<{
+  gql: DebugSpecsFragment
+}>()
+
+const loginConnectStore = useLoginConnectStore()
+
+const run = computed(() => {
+  return props.gql.currentProject?.cloudProject?.__typename === 'CloudProject' && props.gql.currentProject.cloudProject.runByNumber
+})
+
+const debugSpecsArray = computed(() => {
+  if (props.gql.currentProject?.cloudProject?.__typename === 'CloudProject') {
```

Comment:
It sometimes feels like we are building an additional layer in the client, instead of just serving up the data we need. The example I see here is two derived properties based on the fact we *might* have cloud project. Instead, I wonder if we should be doing this on the server, and move the [`specsList`] function there, too. Then client code would become:

```js
const debugSpecs = `computed(() => props.gql.currentProject.cloudProject?.debugSpecs
```

Eg - one of the main things that sold me on GraphQL was the ability to easily add new fields/properties derived from existing data based on changing business requirements (unlike modifying a REST endpoint, which often means needs to modify a lot of other things, too).

Might be more of a philosophical discussion, but the biggest gripe I'm having with the current state of App is the immense complexity of all the computed values based on GraphQL. I wonder if we are doing something wrong? WDYT?

Reply:
I have been having the same feeling as well in a lot of situations, especially around the patterns with the cloud data.  In this instance, the `cloudProject` field is trying to represent four states with the different types.  I am still working with the last remaining IATR M0 issue and still playing around with what to push down to the server layer. I am going to think more about this pattern as it is used elsewhere in the code.  If you have any patterns in mind about these different states, I would love to compare notes.